File: packages/types/src/workflow-rule.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

// Types for the project-scoped workflow rule engine - see
// docs/feature-specs/06-automation-workflow-sla.md ("Moteur de regles
// d'automatisation") in plane-selfhost, and the backend half of this
// feature: apps/api/plane/app/views/workflow_rule/base.py,
// apps/api/plane/app/serializers/workflow_rule.py,
// apps/api/plane/utils/workflow_rule_engine.py. Deliberately named
// `TWorkflowRule*` (not `TRule*`/`TAutomationRule*`) to stay unambiguous
// next to the pre-existing, narrower `TTriageRule*` family
// (packages/types/src/intake/triage-rule.ts), which only fires on intake
// issue creation.

export type TWorkflowTriggerType =
  | "ISSUE_CREATED"
  | "ISSUE_UPDATED"
  | "STATE_CHANGED"
  | "ASSIGNEE_CHANGED"
  | "PRIORITY_CHANGED"
  | "LABEL_ADDED"
  | "COMMENT_ADDED";

/** Only meaningful for `STATE_CHANGED` - both keys optional, either or both
 * may be present. Absent entirely for every other trigger type. */
export type TWorkflowRuleTriggerConfig = {
  from_state_id?: string;
  to_state_id?: string;
};

export type TWorkflowConditionField = "state_id" | "priority" | "label_id" | "assignee_id" | "module_id" | "cycle_id";

export type TWorkflowConditionOperator = "is" | "in" | "is_not" | "not_in" | "is_empty" | "is_not_empty";

export type TWorkflowRuleCondition = {
  field: TWorkflowConditionField;
  operator: TWorkflowConditionOperator;
  /** scalar for `is`/`is_not`, array for `in`/`not_in`, omitted/null for
   * `is_empty`/`is_not_empty`. */
  value?: string | string[] | null;
};

export type TWorkflowActionType =
  | "SET_STATE"
  | "SET_PRIORITY"
  | "SET_ASSIGNEES"
  | "ADD_LABELS"
  | "REMOVE_LABELS"
  | "SET_DUE_DATE"
  | "SET_START_DATE"
  | "POST_COMMENT"
  | "MENTION_USER";

export type TWorkflowActionConfigSetState = { state_id: string };
export type TWorkflowActionConfigSetPriority = { priority: "urgent" | "high" | "medium" | "low" | "none" };
export type TWorkflowActionConfigSetAssignees = { assignee_ids: string[]; mode: "replace" | "add" };
export type TWorkflowActionConfigLabels = { label_ids: string[] };
export type TWorkflowActionConfigDate =
  | { mode: "fixed"; date: string }
  | { mode: "relative"; days_from_trigger: number };
export type TWorkflowActionConfigPostComment = { comment_template: string };
export type TWorkflowActionConfigMentionUser = { user_id: string; comment_template: string };

export type TWorkflowActionConfig =
  | TWorkflowActionConfigSetState
  | TWorkflowActionConfigSetPriority
  | TWorkflowActionConfigSetAssignees
  | TWorkflowActionConfigLabels
  | TWorkflowActionConfigDate
  | TWorkflowActionConfigPostComment
  | TWorkflowActionConfigMentionUser
  | Record<string, never>;

export type TWorkflowAction = {
  id?: string;
  action_type: TWorkflowActionType;
  action_config: TWorkflowActionConfig;
  sort_order: number;
};

export type TWorkflowRule = {
  id: string;
  workspace_id: string;
  project_id: string;
  name: string;
  description: string;
  is_active: boolean;
  trigger_type: TWorkflowTriggerType;
  trigger_config: TWorkflowRuleTriggerConfig;
  conditions: TWorkflowRuleCondition[];
  execution_count: number;
  last_triggered_at: string | null;
  actions: TWorkflowAction[];
  created_at: string;
  updated_at: string;
  created_by: string | null;
};

export type TWorkflowRulePayload = Partial<
  Pick<
    TWorkflowRule,
    "name" | "description" | "is_active" | "trigger_type" | "trigger_config" | "conditions" | "actions"
  >
>;

export type TWorkflowExecutionStatus = "SUCCESS" | "FAILED" | "SKIPPED";

/** One of these four fixed, snake_case reasons whenever `status` is
 * `SKIPPED` - see `evaluate_rules_for_issue` in
 * apps/api/plane/utils/workflow_rule_engine.py. */
export type TWorkflowExecutionSkipReason =
  | "trigger_config_not_matched"
  | "conditions_not_matched"
  | "chain_depth_exceeded"
  | "rate_limit_exceeded";

export type TWorkflowActionApplied = {
  action_id: string;
  action_type: string;
  status: "applied" | "failed";
  detail?: string;
  error?: string;
  comment_id?: string;
};

export type TWorkflowRuleExecutionLog = {
  id: string;
  rule: string;
  issue: string;
  trigger_event: string;
  status: TWorkflowExecutionStatus;
  actions_applied: TWorkflowActionApplied[];
  error_message: string | null;
  chain_depth: number;
  executed_at: string;
};

export type TWorkflowRuleExecutionLogParams = {
  cursor?: string;
  per_page?: number;
  status?: TWorkflowExecutionStatus;
  date_from?: string;
  date_to?: string;
  order_by?: string;
};

File: apps/api/plane/db/models/workflow_rule.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Django imports
from django.db import models

# Module imports
from plane.db.models.base import BaseModel
from plane.db.models.project import ProjectBaseModel


class WorkflowRule(ProjectBaseModel):
    """
    Project-scoped automation rule: trigger + optional AND-combined
    conditions + ordered actions, executed asynchronously on issue events -
    see docs/feature-specs/06-automation-workflow-sla.md ("Moteur de regles
    d'automatisation") in plane-selfhost.

    Structurally similar to TriageRule (`triage_rule.py`), which is
    deliberately left untouched - it only fires at intake-creation time for
    a narrower use case. This is a separate, genuinely new set of tables
    covering the full issue-lifecycle trigger vocabulary below.
    """

    TRIGGER_TYPE_CHOICES = (
        ("ISSUE_CREATED", "Issue created"),
        ("ISSUE_UPDATED", "Issue updated"),
        ("STATE_CHANGED", "State changed"),
        ("ASSIGNEE_CHANGED", "Assignee changed"),
        ("PRIORITY_CHANGED", "Priority changed"),
        ("LABEL_ADDED", "Label added"),
        ("COMMENT_ADDED", "Comment added"),
    )

    name = models.CharField(max_length=255)
    description = models.TextField(blank=True)
    is_active = models.BooleanField(default=True)
    trigger_type = models.CharField(max_length=30, choices=TRIGGER_TYPE_CHOICES)
    # Optional/partial trigger-level filter evaluated against the trigger's
    # own snapshot before conditions are even evaluated - e.g. for
    # STATE_CHANGED: {"from_state_id": "...", "to_state_id": "..."} - either
    # key alone means "any -> X" or "X -> any".
    trigger_config = models.JSONField(default=dict, blank=True)
    # AND-combined leaf conditions, e.g.
    # [{"field": "priority", "operator": "in", "value": ["urgent", "high"]}] -
    # translated to IssueFilterSet leaf dicts at evaluation time, see
    # plane/utils/workflow_rule_engine.py.
    conditions = models.JSONField(default=list, blank=True)
    execution_count = models.IntegerField(default=0)
    last_triggered_at = models.DateTimeField(null=True, blank=True)

    class Meta:
        verbose_name = "Workflow Rule"
        verbose_name_plural = "Workflow Rules"
        db_table = "workflow_rules"
        ordering = ("-created_at",)

    def __str__(self):
        return f"{self.name} <{self.project_id}>"


class WorkflowAction(BaseModel):
    ACTION_TYPE_CHOICES = (
        ("SET_STATE", "Set state"),
        ("SET_PRIORITY", "Set priority"),
        ("SET_ASSIGNEES", "Set assignees"),
        ("ADD_LABELS", "Add labels"),
        ("REMOVE_LABELS", "Remove labels"),
        ("SET_DUE_DATE", "Set due date"),
        ("SET_START_DATE", "Set start date"),
        ("POST_COMMENT", "Post comment"),
        ("MENTION_USER", "Mention user"),
    )

    rule = models.ForeignKey(WorkflowRule, on_delete=models.CASCADE, related_name="actions")
    action_type = models.CharField(max_length=30, choices=ACTION_TYPE_CHOICES)
    # Shapes documented on WorkflowRule's docstring / feature spec, e.g.
    # SET_STATE: {"state_id": "..."} ; SET_ASSIGNEES: {"assignee_ids": [...], "mode": "replace"|"add"} ;
    # POST_COMMENT: {"comment_template": "text with {{issue.identifier}} tokens"}.
    action_config = models.JSONField(default=dict, blank=True)
    sort_order = models.PositiveIntegerField()

    class Meta:
        verbose_name = "Workflow Action"
        verbose_name_plural = "Workflow Actions"
        db_table = "workflow_actions"
        ordering = ("sort_order",)

    def __str__(self):
        return f"{self.action_type} <{self.rule_id}>"


class WorkflowRuleExecutionLog(BaseModel):
    STATUS_CHOICES = (
        ("SUCCESS", "Success"),
        ("FAILED", "Failed"),
        ("SKIPPED", "Skipped"),
    )

    rule = models.ForeignKey(WorkflowRule, on_delete=models.CASCADE, related_name="execution_logs")
    issue = models.ForeignKey(
        "db.Issue", on_delete=models.CASCADE, db_index=True, related_name="workflow_rule_execution_logs"
    )
    # Snapshot of the rule's trigger_type at the time this row was written -
    # a rule's trigger_type could theoretically be edited later, this keeps
    # the historical record accurate.
    trigger_event = models.CharField(max_length=30)
    status = models.CharField(max_length=10, choices=STATUS_CHOICES)
    # Best-effort, sequential application - see
    # plane/utils/workflow_rule_engine.py. Snapshot of what actually
    # happened per action: [{"action_id": "...", "action_type": "...",
    # "status": "applied"|"failed", ...}, ...]
    actions_applied = models.JSONField(default=list, blank=True)
    error_message = models.TextField(null=True, blank=True)
    chain_depth = models.PositiveIntegerField(default=0)
    executed_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        verbose_name = "Workflow Rule Execution Log"
        verbose_name_plural = "Workflow Rule Execution Logs"
        db_table = "workflow_rule_execution_logs"
        ordering = ("-executed_at",)
        indexes = [models.Index(fields=["rule", "executed_at"], name="wf_rule_log_rule_exec_idx")]

    def __str__(self):
        return f"{self.trigger_event} {self.status} <{self.rule_id}>"
